File: attention_sampler/attention_sampler.cpp
template <typename T>
MOBULA_KERNEL cumsum(T* X, T* I, const int N, const int att_size) {
    parfor(N, [&](int i) {
        T* Xi = X + i * att_size;
        T* Ii = I + i * att_size;
        for (int i = 1; i < att_size; ++i) {
            Ii[i] = Ii[i - 1] + Xi[i];
        }
    });
}

template <typename T>
MOBULA_KERNEL map_step(T* attxi, T* index_x, const int N, const float stepx, const float att_size, const int out_size) {
    T* myscale = 2.0 / (att_size - 1);
    parfor(N, [&](int b) {
        int i = 0;
        int j = 0;
        T* mapxi = attxi + b * att_size;
        T* index_i = index_x + b * att_size;
        while (i < out_size)
        {
            if (mapxi[j] >= i*stepx)
            {
                index_i[i] = (j + (i*stepx - mapxi[j]) / (mapxi[j] - mapxi[j - 1])) * myscale - 1.0;
                i++;
            }
            else {
                j++;
            }
        }
    });
}

File: attention_sampler/attention_sampler.py
import mobula


@mobula.op.register
class AttSamplerGrid:
    def __init__(self, scale=1.0, dense=4, iters=5):
        self.scale = scale
        self.dense = dense
        self.iters = iters

    def forward(self, data, attx, atty):
        F = self.F._mobula_hack
        # attx: (N, W, 1)
        # atty: (N, H, 1)
        N, _, in_size, _ = data.shape
        att_size = attx.shape[1]
        out_size = int(in_size * self.scale)
        threshold = float(self.scale * self.dense * in_size) / att_size
        attx = attx * out_size
        atty = atty * out_size
        for j in range(self.iters):
            max_attx = F.max(attx, 1, keepdims=True)  # (N, 1, 1)
            max_atty = F.max(atty, 1, keepdims=True)  # (N, 1, 1)
            if j == 0:
                threshold = F.minimum(F.minimum(
                    max_attx, max_atty), threshold)  # (N, 1, 1)
            else:
                threshold = F.minimum(max_attx, max_atty)
            F.broadcast_minimum(threshold, attx, out=attx)
            F.broadcast_minimum(threshold, atty, out=atty)
            sum_x = F.sum(attx, 1, keepdims=True)  # (N, 1, 1)
            sum_y = F.sum(atty, 1, keepdims=True)  # (N, 1, 1)
            deltax = (out_size - sum_x) / att_size
            deltay = (out_size - sum_y) / att_size
            # compensate the drop value
            attx += deltax
            atty += deltay
        '''
        it is the same as the original implemenation.
        the first element is 1.
        '''
        attx[:, 0] = 1
        atty[:, 0] = 1
        if hasattr(F, 'cumsum'):
            attxi = F.cumsum(attx, 1)
            attyi = F.cumsum(atty, 1)
        else:
            attxi = self.F.empty((N, att_size, 1))
            attyi = self.F.empty((N, att_size, 1))
            attxi[:, 0] = 1
            attyi[:, 0] = 1
            mobula.func.cumsum(N, attx, attxi, att_size)
            mobula.func.cumsum(N, atty, attyi, att_size)
        stepx = attxi[:, -1] / out_size
        stepy = attyi[:, -1] / out_size
        ctx = F.get_ctx(stepx)
        index_x = F.empty((N, out_size, 1), ctx=ctx)
        index_y = F.empty((N, out_size, 1), ctx=ctx)
        # note: x and y are inverse.
        mobula.func.map_step(N, attxi, index_y, stepx, att_size, out_size)
        mobula.func.map_step(N, attyi, index_x, stepy, att_size, out_size)
        return F.tile(F.reshape(index_x, (N, 1, out_size)), (1, out_size, 1)),\
            F.tile(index_y, (1, 1, out_size))

    def backward(self, dy_x, dy_y):
        return [0, 0, 0]

    def infer_shape(self, in_shape):
        dshape = in_shape[0]
        out_size = int(dshape[2] * self.scale)
        oshape = (dshape[0], out_size, out_size)
        return in_shape, [oshape, oshape]
